File: 5.最长回文子串.js
/*
 * @lc app=leetcode.cn id=5 lang=javascript
 *
 * [5] 最长回文子串
 */
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    let s_length = s.length;
    let sub_s_index = 0;

    while (s_length >= 1) {
        while (sub_s_index + s_length <= s.length) {
            let pal_str = s.substr(sub_s_index, s_length);
            if (isPalindrome(pal_str)) {
                return pal_str;
            } else {
                sub_s_index++;
            }
        }

        s_length--;
        sub_s_index = 0;
    }

    return '';
};

//判断回文
isPalindrome = function(str) {
    while (str.length >= 0) {
        if (str.length <= 1) return true;

        if (str[0] === str[str.length - 1]) {
            str = str.slice(1, -1);
        } else {
            return false;
        }
    }

    return false;
};

File: 5.最长回文子串.py
#
# @lc app=leetcode.cn id=5 lang=python3
#
# [5] 最长回文子串
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        # 特例
        slen = len(s)
        if slen < 2:
            return s
        maxLen = 1
        begin = 0
        dp = [[-1 for i in range(slen)] for j in range(slen)]
        charArr = list(s)
        for i in range(0, slen):
            dp[i][i] = 0
        for j in range(1, slen):
            for i in range(0, j):
                if charArr[i] != charArr[j]:
                    dp[i][j] = False
                else:
                    if j-i < 3:
                        dp[i][j] = True
                    else:
                        dp[i][j] = dp[i+1][j-1]
                if dp[i][j] and j - i + 1 > maxLen:
                    maxLen = j - i + 1
                    begin = i
        return s[begin:begin + maxLen]
# @lc code=end


solution = Solution()
res = solution.longestPalindrome('babad')
print(res)
